source/slang/slang-ir-legalize-varying-params.cpp

```diff
@@ -177,6 +177,118 @@ void assign(IRBuilder& builder, LegalizedVaryingVal const& dest, IRInst* src)
     assign(builder, dest, LegalizedVaryingVal::makeValue(src));
 }
 
+
+// Several of the derived calcluations rely on having
+// access to the "group extents" of a compute shader.
+// That information is expected to be present on
+// the entry point as a `[numthreads(...)]` attribute,
+// and we define a convenience routine for accessing
+// that information.
+
+IRInst* emitCalcGroupExtents(
+    IRBuilder& builder,
+    IRFunc* entryPoint,
+    IRVectorType* type)
+{
+    if (auto numThreadsDecor = entryPoint->findDecoration<IRNumThreadsDecoration>())
+    {
+        static const int kAxisCount = 3;
+        IRInst* groupExtentAlongAxis[kAxisCount] = {};
+
+        for (int axis = 0; axis < kAxisCount; axis++)
+        {
+            auto litValue = as<IRIntLit>(numThreadsDecor->getExtentAlongAxis(axis));
+            if (!litValue)
+                return nullptr;
+
+            groupExtentAlongAxis[axis] = builder.getIntValue(type->getElementType(), litValue->getValue());
+        }
+
+        return builder.emitMakeVector(type, kAxisCount, groupExtentAlongAxis);
+    }
+
+    // TODO: We may want to implement a backup option here,
+    // in case we ever want to support compute shaders with
+    // dynamic/flexible group size on targets that allow it.
+    //
+    SLANG_UNEXPECTED("Expected '[numthreads(...)]' attribute on compute entry point.");
+    UNREACHABLE_RETURN(nullptr);
+}
+
+// There are some cases of system-value inputs that can be derived
+// from other inputs; notably compute shaders support `SV_DispatchThreadID`
+// and `SV_GroupIndex` which can both be derived from the more primitive
+// `SV_GroupID` and `SV_GroupThreadID`, together with the extents
+// of the thread group (which are specified with `[numthreads(...)]`).
+//
+// As a utilty to target-specific subtypes, we define helpers for
+// calculating the value of these derived system values from the
+// more primitive ones.
+
+    /// Emit code to calculate `SV_DispatchThreadID`
+IRInst* emitCalcDispatchThreadID(
+    IRBuilder& builder,
+    IRType* type,
+    IRInst* groupID,
+    IRInst* groupThreadID,
+    IRInst* groupExtents)
+{
+    // The dispatch thread ID can be computed as:
+    //
+    //      dispatchThreadID = groupID*groupExtents + groupThreadID
+    //
+    // where `groupExtents` is the X,Y,Z extents of
+    // each thread group in threads (as given by
+    // `[numthreads(X,Y,Z)]`).
+
+    return builder.emitAdd(type,
+        builder.emitMul(type,
+            groupID,
+            groupExtents),
+        groupThreadID);
+}
```

Comment:
I can't move this function back currently (may change due to other review comments and redesign of code).

I can't move it since `emitCalcGroupExtents` was originally inside a class which is not supposed to be available beyond the current compilation unit.